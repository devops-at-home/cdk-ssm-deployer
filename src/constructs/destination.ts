import { Construct } from 'constructs';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import { Rule } from 'aws-cdk-lib/aws-events';
import { AwsApi } from 'aws-cdk-lib/aws-events-targets';
import { PolicyStatement } from 'aws-cdk-lib/aws-iam';

type SSMDestinationProps = {
  bucketName: string;
  destination: string;
};

export class SSMDestination extends Construct {
  constructor(scope: Construct, id: string, props: SSMDestinationProps) {
    super(scope, id);

    const { bucketName, destination } = props;

    new NodejsFunction(this, 'Lambda', {
      functionName: `deployer-${destination}-get-current-version`,
    });

    const s3BucketRule = new Rule(this, 'Rule', {
      ruleName: `deployTrigger-${destination}`,
      eventPattern: {
        source: ['aws.s3'],
        detail: {
          eventSource: ['s3.amazonaws.com'],
          eventName: ['PutObject'],
          requestParameters: {
            bucketName,
            key: `${destination}.zip`,
          },
        },
      },
    });

    const policyStatement = new PolicyStatement({
      actions: ['ssm:*'],
      resources: ['*'],
    });

    const ssmApiAction = new AwsApi({
      action: 'action',
      service: 'service',
      apiVersion: 'apiVersion',
      catchErrorPattern: 'catchErrorPattern',
      parameters: { destination },
      policyStatement,
    });

    s3BucketRule.addTarget(ssmApiAction);

    // TODO: OIDC permissions for git repo
    // - write to bucket folder (zip only)
    // - invoke version lambda (is it possible to restrict input params) -> only way to do this is a lambda per destination

    // TODO: ssm role and permissions to
    // - read from bucket folder
    // - invoke version lambda
    // - trigger deployment

    // TODO: custom resource to create destination folder in bucket
  }
}
