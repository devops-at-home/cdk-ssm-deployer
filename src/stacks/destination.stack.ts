import { Stack, StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { SSMRole, SSMRoleConfig } from '../constructs/ssm-role.construct';

interface DestinationStackProps extends SSMRoleConfig, StackProps {
    instanceName: string;
    environment: string;
}

export class DestinationStack extends Stack {
    public readonly roleName: string;
    constructor(scope: Construct, id: string, props: DestinationStackProps) {
        super(scope, id, props);

        const { roleName } = new SSMRole(this, 'Role', {
            ...props,
        });

        this.roleName = roleName;
    }
}
