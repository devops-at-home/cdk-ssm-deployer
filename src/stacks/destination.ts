import { NestedStack, NestedStackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { SSMRole } from '../constructs/ssm-role';
// import { SSMDeployment } from '../constructs/ssm-deployment';

interface DestinationStackProps extends NestedStackProps {
  bucketName: string;
  tableName: string;
  destination: string;
  keyArn: string;
}

export class DestinationStack extends NestedStack {
  constructor(scope: Construct, id: string, props: DestinationStackProps) {
    super(scope, id, props);

    const { bucketName, tableName, destination, keyArn } = props;

    new SSMRole(this, 'Role', {
      destination,
      bucketName,
      tableName,
      keyArn,
    });

    // new SSMDeployment(this, 'Deployment', {
    //   bucketName,
    //   destination,
    // });
  }
}
