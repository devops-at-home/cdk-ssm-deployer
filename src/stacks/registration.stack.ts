import { Stack, StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { SSMRegistration, SSMRegistrationProps } from '../constructs/ssm-registration.construct';

interface RegistrationStackProps extends StackProps, SSMRegistrationProps {}

export class RegistrationStack extends Stack {
    constructor(scope: Construct, id: string, props: RegistrationStackProps) {
        super(scope, id, props);

        const { instanceName, roleName } = props;

        new SSMRegistration(this, 'SSMRegistration', {
            instanceName,
            roleName,
        });
    }
}
