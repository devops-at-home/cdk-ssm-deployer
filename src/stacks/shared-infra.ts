import { NestedStack, NestedStackProps } from 'aws-cdk-lib';
import { Table } from 'aws-cdk-lib/aws-dynamodb';
import { Bucket } from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';
import { BucketWithEventBridge } from '../constructs/config-bucket';
import { VaultDynamoDB } from '../constructs/vault-dynamodb';
import { GitHubActionsOidcProvider } from '../constructs/gh-aws-oidc-connect-provider';
import { GitHubActionsRole } from '../constructs/gh-aws-oidc-connect-role';
import { PolicyDocument, PolicyStatement } from 'aws-cdk-lib/aws-iam';
import { KmsKey } from '../constructs/kms-key';
import { Key } from 'aws-cdk-lib/aws-kms';

const repository = 'devops-at-home/cdk-ssm-deployer';
const destinations: string[] = [];

export class SharedInfraStack extends NestedStack {
  public bucket: Bucket;
  public table: Table;
  public key: Key;
  constructor(scope: Construct, id: string, props: NestedStackProps = {}) {
    super(scope, id, props);

    // TODO: SSM document for running deployments

    const { bucket } = new BucketWithEventBridge(this, 'Bucket');
    this.bucket = bucket;

    const { table } = new VaultDynamoDB(this, 'VaultDynamoDB');
    this.table = table;

    const { key } = new KmsKey(this, 'KmsKey');
    this.key = key;

    new GitHubActionsOidcProvider(this, 'GitHubActionsOidcProvider');

    new GitHubActionsRole(this, 'GitHubActionsRole', {
      provider: GitHubActionsOidcProvider.forAccount(),
      repository,
      inlinePolicies: {
        s3: new PolicyDocument({
          statements: [
            new PolicyStatement({
              resources: [`${bucket.bucketArn}/*`],
              actions: ['s3:PutObject'],
              conditions: {
                StringEquals: {
                  's3:prefix': destZips(destinations),
                },
              },
            }),
          ],
        }),
      },
    });
  }
}

const destZips = (destinations: string[]) => {
  return destinations.map((dest) => {
    return `${dest}.zip`;
  });
};
