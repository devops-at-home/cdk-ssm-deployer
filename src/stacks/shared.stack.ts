import { CfnOutput, Stack, StackProps } from 'aws-cdk-lib';
import { Bucket, BucketEncryption, IBucket } from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';

interface SharedStackProps extends StackProps {
    environment: string;
}

export class SharedStack extends Stack {
    public readonly bucket: IBucket;
    constructor(scope: Construct, id: string, props: SharedStackProps) {
        super(scope, id, props);

        const { environment } = props;

        const { bucketArn } = (this.bucket = new Bucket(this, 'Bucket', {
            encryption: BucketEncryption.S3_MANAGED,
        }));

        new CfnOutput(this, 'CfnOutput', {
            exportName: getBucketExportName(environment),
            value: bucketArn,
        });
    }
}

export const getBucketExportName = (environment: string) => {
    return `app-config-backup-bucket-${environment}`;
};
