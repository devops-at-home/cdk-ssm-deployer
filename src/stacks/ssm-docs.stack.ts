import { Stack, StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';

interface SSMDocsStackProps extends StackProps {}

export class SSMDocsStack extends Stack {
    constructor(scope: Construct, id: string, props: SSMDocsStackProps) {
        super(scope, id, props);

        // TODO: SSM document for running deployments
    }
}
