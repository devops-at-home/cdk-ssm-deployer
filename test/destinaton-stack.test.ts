import { Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { DestinationStack } from '../src/stacks/destination';

describe('DestinationStack', () => {
  const stack = new Stack();
  const nestedStack = new DestinationStack(stack, 'DestinationStack', {
    bucketName: 'bucketName',
    tableName: 'tableName',
    destination: 'h6060-001',
  });

  const template = Template.fromStack(nestedStack);

  test('Snapshot', () => {
    expect(template.toJSON()).toMatchSnapshot();
  });
});
