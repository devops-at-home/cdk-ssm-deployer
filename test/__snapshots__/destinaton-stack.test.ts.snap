// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DestinationStack Snapshot 1`] = `
Object {
  "Mappings": Object {
    "ServiceprincipalMap": Object {
      "af-south-1": Object {
        "ssm": "ssm.af-south-1.amazonaws.com",
      },
      "ap-east-1": Object {
        "ssm": "ssm.ap-east-1.amazonaws.com",
      },
      "ap-northeast-1": Object {
        "ssm": "ssm.amazonaws.com",
      },
      "ap-northeast-2": Object {
        "ssm": "ssm.amazonaws.com",
      },
      "ap-northeast-3": Object {
        "ssm": "ssm.amazonaws.com",
      },
      "ap-south-1": Object {
        "ssm": "ssm.amazonaws.com",
      },
      "ap-southeast-1": Object {
        "ssm": "ssm.amazonaws.com",
      },
      "ap-southeast-2": Object {
        "ssm": "ssm.amazonaws.com",
      },
      "ap-southeast-3": Object {
        "ssm": "ssm.ap-southeast-3.amazonaws.com",
      },
      "ca-central-1": Object {
        "ssm": "ssm.amazonaws.com",
      },
      "cn-north-1": Object {
        "ssm": "ssm.amazonaws.com",
      },
      "cn-northwest-1": Object {
        "ssm": "ssm.amazonaws.com",
      },
      "eu-central-1": Object {
        "ssm": "ssm.amazonaws.com",
      },
      "eu-north-1": Object {
        "ssm": "ssm.amazonaws.com",
      },
      "eu-south-1": Object {
        "ssm": "ssm.eu-south-1.amazonaws.com",
      },
      "eu-south-2": Object {
        "ssm": "ssm.eu-south-2.amazonaws.com",
      },
      "eu-west-1": Object {
        "ssm": "ssm.amazonaws.com",
      },
      "eu-west-2": Object {
        "ssm": "ssm.amazonaws.com",
      },
      "eu-west-3": Object {
        "ssm": "ssm.amazonaws.com",
      },
      "me-south-1": Object {
        "ssm": "ssm.me-south-1.amazonaws.com",
      },
      "sa-east-1": Object {
        "ssm": "ssm.amazonaws.com",
      },
      "us-east-1": Object {
        "ssm": "ssm.amazonaws.com",
      },
      "us-east-2": Object {
        "ssm": "ssm.amazonaws.com",
      },
      "us-gov-east-1": Object {
        "ssm": "ssm.amazonaws.com",
      },
      "us-gov-west-1": Object {
        "ssm": "ssm.amazonaws.com",
      },
      "us-iso-east-1": Object {
        "ssm": "ssm.amazonaws.com",
      },
      "us-iso-west-1": Object {
        "ssm": "ssm.us-iso-west-1.amazonaws.com",
      },
      "us-isob-east-1": Object {
        "ssm": "ssm.amazonaws.com",
      },
      "us-west-1": Object {
        "ssm": "ssm.amazonaws.com",
      },
      "us-west-2": Object {
        "ssm": "ssm.amazonaws.com",
      },
    },
  },
  "Resources": Object {
    "Role8C6B06EB": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Object {
                  "Fn::FindInMap": Array [
                    "ServiceprincipalMap",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    "ssm",
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/AmazonSSMManagedInstanceCore",
              ],
            ],
          },
        ],
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": "s3:GetObject",
                  "Effect": "Allow",
                  "Resource": "arn:aws:s3:::bucketName/h6060-001.zip",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "bucket",
          },
        ],
        "RoleName": "SSMServiceRole-h6060-001",
      },
      "Type": "AWS::IAM::Role",
    },
    "RoleTablePolicyED9E5ADA": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "dynamodb:DescribeLimits",
                "dynamodb:DescribeTimeToLive",
                "dynamodb:ListTagsOfResource",
                "dynamodb:DescribeReservedCapacityOfferings",
                "dynamodb:DescribeReservedCapacity",
                "dynamodb:BatchGetItem",
                "dynamodb:BatchWriteItem",
                "dynamodb:CreateTable",
                "dynamodb:DeleteItem",
                "dynamodb:GetItem",
                "dynamodb:GetRecords",
                "dynamodb:PutItem",
                "dynamodb:Query",
                "dynamodb:UpdateItem",
                "dynamodb:Scan",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:dynamodb:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":table/tableName",
                  ],
                ],
              },
            },
            Object {
              "Action": "dynamodb:ListTables",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:dynamodb:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":table/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "table",
        "Roles": Array [
          Object {
            "Ref": "Role8C6B06EB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
